lib/puppet/pops/pcore.rb

```diff
@@ -42,6 +42,16 @@ def self.init(loader, ir, for_agent)
     if Puppet[:tasks]
       require_relative 'types/task'
       require_relative 'types/execution_result'
+
+      add_object_type('Pcore::Node', <<-PUPPET, loader)
+      {
+        init => { uri => Init[URI] },
+        attributes => {
+          uri => URI,
+          options => { type => Hash[String[1], Data], value => {} }
+        }
```

Comment:
I suspect that it is also wanted to be able to initialize a Node from a hash with keys `uri` and `options`. Not sure what the `init => uri => Init[URI]` thing means - can you elaborate?

Reply:
It is already possible to initialize any `Object` with a hash keyed by the attribute names. Two constructors are always generated, one that takes positional arguments and one that takes a named arguments (i.e. the hash that you ask for here).

I needed a way to create a `Node` object from a `String` although its required attribute is an `URI`. Obviously that didn't work since both generated constructors expected the type to be `URI`. The real problem isn't about strings and uris though, but about passing arguments that are instances of the `Init[T]` type for the corresponding attribute type `T`. I was torn between the following alternatives:

1. Always allow that all parameters can be passed as Init[T] instead of T and then auto-converted.
2. Like #<span>1</span> but limit this new behavior to constructors (i.e. the #new function).
3. Like #<span>2</span> but limit to `Object` constructors.
4. Explicitly control this behavior by adding an extra entry to `Object`.

The first approach would create havoc among dispatchers all over Puppet since many of them rely on (and alter behavior based on) the fact  that it does not work that way. The same is true for #<span>2</span>. In the choice between #<span>3</span> and #<span>4</span>, I chose the latter because I felt that a declaration was needed to remove magic. I then tested different approaches with the objective to be able to modify the generated constructors (using annotations was one of them) but finally realized two things. a) An `Object` type should not aspire to describe the implementation (we have other plans for that) and b) the only thing needed here is to describe that some attributes should be able to be initialized using `Init`. The absolute simplest way to declare this, is to add:
```
init => { uri => Init[URI] }
```
which states "the constructor (the init method) will accept a value for the 'uri' attribute that can be coerced using `Init[URI]`".

Reply:
Why then as a separate thing instead of being an attribute of an attribute - i.e. write something like this:
```
attributes { uri => { type => URI, init => Init[URI] } }
```

Reply:
Yes, it could be written like that.